tests/unit/models/gpflux/test_sampler.py

```diff
@@ -0,0 +1,159 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+In this module, we test the *behaviour* of Trieste models against reference GPflux models (thus
+implicitly assuming the latter are correct).
+*NOTE:* Where GPflux models are used as the underlying model in an Trieste model, we should
+*not* test that the underlying model is used in any particular way. To do so would break
+encapsulation. For example, we should *not* test that methods on the GPflux models are called
+(except in the rare case that such behaviour is an explicitly documented behaviour of the
+Trieste model).
+"""
+
+from __future__ import annotations
+
+from typing import Callable, Tuple
+
+import numpy.testing as npt
+import pytest
+import tensorflow as tf
+from gpflux.models import DeepGP
+
+from tests.util.misc import TF_DEBUGGING_ERROR_TYPES, ShapeLike, mk_dataset, quadratic, random_seed
+from tests.util.models.gpflow.models import QuadraticMeanAndRBFKernel
+from tests.util.models.gpflux.models import two_layer_trieste_dgp
+from trieste.data import Dataset
+from trieste.models.gpflux import DeepGaussianProcess
+from trieste.models.gpflux.sampler import DeepGaussianProcessReparamSampler
+from trieste.space import Box
+from trieste.types import TensorType
+
+
+@pytest.mark.parametrize("sample_size", [0, -2])
+def test_deep_gaussian_process_sampler_raises_for_invalid_sample_size(
+    sample_size: int, keras_float: None
+) -> None:
+    search_space = Box([0.0], [1.0]) ** 4
+    x = search_space.sample(10)
+    data = mk_dataset(x, quadratic(x))
+
+    dgp = two_layer_trieste_dgp(data, search_space)
+
+    with pytest.raises(TF_DEBUGGING_ERROR_TYPES):
+        DeepGaussianProcessReparamSampler(sample_size, dgp)
+
+
+def test_deep_gaussian_process_sampler_raises_for_invalid_model() -> None:
+    with pytest.raises(ValueError, match="Model must be .*"):
+        DeepGaussianProcessReparamSampler(10, QuadraticMeanAndRBFKernel())  # type: ignore
+
+
+@pytest.mark.parametrize("shape", [[], [1], [2], [2, 3, 4]])
+def test_deep_gaussian_process_sampler_sample_raises_for_invalid_at_shape(
+    shape: ShapeLike,
+    keras_float: None,
+) -> None:
+    search_space = Box([0.0], [1.0])
+    x = search_space.sample(10)
+    data = mk_dataset(x, quadratic(x))
+
+    dgp = two_layer_trieste_dgp(data, search_space)
+    sampler = DeepGaussianProcessReparamSampler(1, dgp)
+
+    with pytest.raises(TF_DEBUGGING_ERROR_TYPES):
+        sampler.sample(tf.zeros(shape))
+
+
+def build_dataset_and_train_deep_gp(
```

Comment:
if used only here make it a private function?
```suggestion
def _build_dataset_and_train_deep_gp(
```